src/metrics/rules/active_ruleset.go

```diff
@@ -45,6 +48,69 @@ type activeRuleSet struct {
 	tagsFilterOpts  filters.TagsFilterOptions
 	newRollupIDFn   metricid.NewIDFn
 	isRollupIDFn    metricid.MatchIDFn
+
+	useFastMatch bool
+	// The key is the value of the __name__ label, the value is the mappingRule
+	mappingRuleIndex map[string][]*mappingRule
+	// The key is the value of the __name__ label, the value is rollupRule
+	rollupRuleIndex map[string][]*rollupRule
+}
+
+func newActiveRuleSetFast(
+	version int,
+	mappingRules []*mappingRule,
+	rollupRules []*rollupRule,
+	tagsFilterOpts filters.TagsFilterOptions,
+	newRollupIDFn metricid.NewIDFn,
+	isRollupIDFn metricid.MatchIDFn,
+) *activeRuleSet {
+	rset := newActiveRuleSet(version, mappingRules, rollupRules, tagsFilterOpts, newRollupIDFn, isRollupIDFn)
+	rset.useFastMatch = true
+
+	nowNanos := time.Now().UnixNano()
+
+	getNameTagValue := func(nameFilterValue *filters.FilterValue, snapshotString func() string) string {
+		if nameFilterValue == nil {
+			panic(fmt.Sprintf("A rule has no fitler for metric name tag: %s", snapshotString()))
+		}
+		if nameFilterValue.Negate {
+			panic(fmt.Sprintf("A rule has negated filter for metric name tag: %s", snapshotString()))
+		}
+		if strings.Contains(nameFilterValue.Pattern, "*") {
+			panic(fmt.Sprintf("A rule has wildcard filter for metric name tag %s", snapshotString()))
+		}
+		return nameFilterValue.Pattern
+	}
+
+	rset.mappingRuleIndex = make(map[string][]*mappingRule)
+	for _, rule := range mappingRules {
+		snapshot := rule.activeSnapshot(nowNanos)
+		if snapshot == nil {
+			panic(fmt.Sprintf("mapping rule has no active snapshot: %+v", rule))
+		}
+		nameTagValue := getNameTagValue(snapshot.filter.NameFilterValue(), func() string { return fmt.Sprintf("%+v", snapshot) })
+		rset.mappingRuleIndex[nameTagValue] = append(rset.mappingRuleIndex[nameTagValue], rule)
```

Comment:
IIUC seems the order preserve and if the first mapping rule matches, it will short circuit? fyi @hczhu-db

Reply:
The order is preserved, though the order doesn't matter since a data sample matches against all rules and can get multiple matched rules. Each matched rule (mapping or rollup) is processed independently.  CC @karan-db 
The logic workflow
```
potential_rules  = all rules with the same metric name
for rule in potential_rules {
  do a full matching for all labels
  if !match {
    continue
  }
  // Process the matched rule drop or keep for a mapping rule
  // Send the rule and data sample to the aggregation pipeline if it's a rollup rule.
}

```